Sprint 1

Nilfan opened the pull request.

Commit:
First commit

Add common files .gitignore, .nvmrc, package.json and other config files
Add index.html, index.js
Add login, singin and error pages
Add helpers

Commit:
Add build-script

Commit:
Add netlify config

Commit:
Fix netlify config
Fix build script path

Commit:
Fix netlify config

Commit:
Delete sass config
Fix build command

Commit:
Fix start command

Commit:
Fix build command

Commit:
Fix imports

Commit:
Change netlify config

Commit:
Change static folder

Commit:
Change dirname of web-server

Commit:
Fix class imports
Fix module import in index.html
Add end-point to return index.html every time

Commit:
Fix server.js

Commit:
Fix server

Commit:
Fix server

Commit:
Delete redirect from toml-config

Commit:
Change toml-config

Commit:
Fix server.js
Fix link class

Commit:
Add force redirection [netlify.toml config]

Commit:
Fix redirection settings

Commit:
Fix from path

Commit:
netlify.toml:

Fix from path
Add header in

Commit:
netlify.toml:

Change from param

Commit:
Try multiple from params

Commit:
netlify.toml:

Change from param

Commit:
netlify.toml:

Add redirects

Commit:
Add root path in routes as Login page
Add user profile page
Add ui prototype

Commit:
Add ReadMe

Commit:
Add Netlify link in ReadMe

Commit:
Fix readme

Commit:
Fix links in readme

avdoshenkov commented on .gitignore:
```diff
@@ -0,0 +1,13 @@
+// build
```
В `.gitignore` комментарии начинаются не как в JS, а со знака `#`

Nilfan replied:
Done

avdoshenkov commented on README.md:
```diff
@@ -1,82 +1,25 @@
-### Ветка, в которой делаете задания спринта, должна называться sprint_i, где i - номер спринта. Не переименовывайте её.
+## Usage
```
В самом начале будет круто, если добавишь описание и название проекта — что это вообще за репозиторий и для чего. А потом уже да, можно про запуск рассказать.

Nilfan replied:
ReadMe changed in main branch

avdoshenkov commented on server.js:
```diff
@@ -0,0 +1,19 @@
+const path = require("path");
+
+const app = express();
+const PORT = 3000;
```
Здесь можно порт указать таким образом:
```
const PORT = process.env.PORT || 3000; 
```
Тогда будет возможность брать значение из env переменных, а уже если там её не окажется, то брать значение по умолчанию

Nilfan replied:
Done

avdoshenkov commented on src/pages/login-page/login-page.tmpl.js:
```diff
@@ -0,0 +1,20 @@
+      <form class="column-vertical-align">
+        <h2 class="title"> Вход </h2>
+        <div class="form-field">
+          <input class="text-input invalid" name="login" placeholder="Логин"/>
```
По хорошему у каждого инпута должен быть свой лейбл. Это и с точки зрения семантики, и доступности будет полезно, да и пользователю удобно. Конкретно данный макет, конечно, подталкивает использовать placeholder, но всё же это не лучшее решение. Подробнее можно почитать [тут](https://www.htmhell.dev/24-a-placeholder-is-not-a-label/)

Nilfan replied:
Done

avdoshenkov commented on src/pages/signin-page/signin-page.tmpl.js:
```diff
@@ -0,0 +1,48 @@
+export const signinPageTmpl = `
```
`Sign in` — это всё же имеется ввиду вход обычно, а для регистрации — `Sign up`

Nilfan replied:
Fixed

avdoshenkov commented on src/pages/signin-page/signin-page.tmpl.js:
```diff
@@ -0,0 +1,48 @@
+        </div>
+        <div class="controls"> 
+          <button class="btn btn-ok"> Зарегистрироваться </button>
+          <button class="btn btn-cancel"> Войти </button>
```
Здесь скорее всего по смыслу больше подходит ссылка, т.к. по клику нас должно перекинуть на отдельную страницу со входом. Это достаточно [распространённая штука](https://www.htmhell.dev/5-button-like-link/), которую лучше избегать

Nilfan replied:
Fixed

avdoshenkov commented on src/styles/common-style.scss:
```diff
@@ -0,0 +1,7 @@
+$main-blue: #3587be;
```
Конкретно этот файл, кажется, можно переименовать в `variables.scss`. По смыслу будет логичнее. 

В `common` обычно какие-то классы записывают, которые по системе в разных местах будут использоваться, по типу тех, что уже в `style.scss` определены

Nilfan replied:
File renamed

avdoshenkov commented on src/utils/helpers.js:
```diff
@@ -0,0 +1,162 @@
+  return result;
+}
+
+function identity(param) {
```
Не совсем понял смысл этой функции

Nilfan replied:
Function were deleted

avdoshenkov commented on src/utils/helpers.js:
```diff
@@ -0,0 +1,162 @@
+
+const baseRange = (start, end, step, isRight) => {
+  let index = -1;
+  let length = Math.max(Math.ceil((end - start) / (step || 1)), 0);
```
```suggestion
const baseRange = (start, end, step = 1, isRight) => {
  let index = -1;
  let length = Math.max(Math.ceil((end - start) / step), 0);
```
Значение по умолчанию лучше вынести сюда, т.к. дальше в цикле тоже используется переменная `step` и если она не определена, то результат выполнения будет не тем, что мы ожидаем

avdoshenkov replied:
Имя аргумента `isRight` не совсем очевидное. В данном случае его лучше назвать как-то вроде `hasReverseOrder` или что-то подобное

Nilfan replied:
Argument renamed

avdoshenkov commented on src/utils/helpers.js:
```diff
@@ -0,0 +1,162 @@
+  return baseRange(start, end, step, isRight);
+}
+
+function rangeRight(start, end, step) {
```
Эту функцию тоже лучше назвать по типу `reversedRange` или `descendingRange`

Nilfan replied:
Also renamed

avdoshenkov commented on src/utils/helpers.js:
```diff
@@ -0,0 +1,162 @@
+
+function isIterable(obj) {
+  // checks for null and undefined
+  if (obj == null) {
```
Раз мы проверяем и `undefined`, то можно использовать созданную функцию `isNil`

Nilfan replied:
Done

avdoshenkov commented on src/utils/helpers.js:
```diff
@@ -0,0 +1,162 @@
+  }
+
+  for (const key in value) {
+    if (hasOwnProperty.call(value, key)) {
```
Можно просто:
```suggestion
    if (value.hasOwnProperty(key)) {
```

Nilfan replied:
Suggestion applied

avdoshenkov commented on src/utils/helpers.js:
```diff
@@ -0,0 +1,162 @@
+    (Array.isArray(value) ||
+      typeof value === "string" ||
+      typeof value.splice === "function" ||
+      isBuffer(value) ||
+      isTypedArray(value) ||
```
Эти две функции не определены

Nilfan replied:
Both functions were deleted

avdoshenkov commented on src/pages/user-profile/user-profile.js:
```diff
@@ -0,0 +1,49 @@
+
+  compile(ctx) {
+
+    const raws = this.generateUserInfoRows(ctx)
```
Наверное, имелось в виду `rows`

Nilfan replied:
Fixed

avdoshenkov reviewed (changes requested):
Отлично поработал! Осталось только несколько нюансов поправить, и будет идеально.

Ещё заметил, что по заданию не хватает страницы с мессенджером (хотя бы заглушки), а также полей для смены пароля
<img width="676" alt="Снимок экрана" src="https://user-images.githubusercontent.com/26881016/124829579-d9a64c00-df89-11eb-936d-bb3b314f8dfc.png">

Commit:
Fix .gitignore
Move routes in separate file
Add routes - change-user-profile, change-user-password, messenger
Fix port choosing
Optimize util-functions
Move colors in variables.scss
Add components (aka tables) for user-profile-page - user-info, change-user-info, change-user-password
Add messenger dummy page
Add new design file

avdoshenkov commented on src/components/user-info-table/user-info-table.tmpl.js:
```diff
@@ -0,0 +1,32 @@
+    <div class="two-column-grid user-info-table">
+      {{#each rows}}
+        <div class="grid-raw">
+          <div class="raw-key"> {{this.key}} </div>
+          <div class="raw-value"> {{this.value}} </div>
```
Здесь можно использовать [Description list](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl). Это просто на будущее)

Nilfan replied:
Полезно, запомню)

avdoshenkov reviewed (approved):
Всё супер, молодцом! Можно двигаться дальше

Commit:
Update ReadMe
